### DISCUSSION TITLE
Think about & share your Use Cases

### DISCUSSION THREAD

**Discussion by NaitLee** [NaitLee/Cat-Printer#27] [General]
Such kind of printers are designed for school students, and usually used to print learning materials (aka “错题打印机”), to help improve learning efficiency.

'Till now I see guys have much more use cases, and notably "Live Printing" (Text from chatroom?) and Network Printing (CUPS/IPP).
That means the app/program/driver should be flexible enough.

The goal of this project is obvious: consider more common use cases, and make it flexible enough, without too much.
And best of all, **it's free-as-in-freedom.**

You can also share how you're using the printer & this project, or imagine some. This may inspire more ideas 😁
And I encourage to join the development -- **it's not difficult at all!** Knowing a little about programming is sufficient.

*disclaimer: I have no relationship with any of the printer vendors, nor do I promote any of the products*

**Comment by NaitLee** [comment]
One scenario come to my mind:

- School doesn't allow taking a (smart)phone.
- Students want to do less writing, with this kind of printer, to take notes quickly & focus on *true* knowledge.
- Classroom have Bluetooth-compatible computer. Someone could manage it.
  ... or the teacher's Android phone! If teacher is a kind person.
- Use this project to help! Forget those non-sense "original" apps.

My high-school life had all conditions satisfied -- except for a printer.
For this, automatic "batch-printing" could be useful. Let me know if anyone want this.

**Comment by andypiper** [comment]
I was thinking about the network printing side of it some more, because then I could use any device that can find a printer on my home network instead of using BLE. Just have the printer on BLE to a single machine (say a Linux box like a Pi, or a desktop Mac), and then have Airprint from phones or iPads etc. That would be cool, although I'm not sure where image conversion should happen, it would have to be in the printer server. I might dig into this some more sometime.

**Reply by NaitLee** [reply]
The server is capable to convert PostScript (with Ghostscript). That means other clients could just support PostScript.

In theory other formats could be easily supported too, but I can't see how IPP could tell the data format. We can just guess from file content (magic headers).

Besides that, there are more works needed:

- Security & authentication, maybe also limit PS/GS functionality because PS can do executions/heavy calculations.
- Maybe needs broadcasting the service
- Tell other devices about more specs (format supported, paper width/color etc.)
- And maybe more file/data "format" to support

Learning more about IPP protocol is required, but the [spec doc](https://datatracker.ietf.org/doc/html/rfc8010) is just not for human-beings...
... or pick another protocol?

Other printer-related projects are all printer *clients* (or "central") for controlling printers, while this is a dummy "peripheral" that mimics a physical printer. I haven't seen a (good) later case that could give us examples.

**Reply by wenestvedt** [reply]
This is my interest, too: run CUPS on a Raspberry Pi or Ubuntu box, listening for network print jobs from phones and laptops around the house.

I have a family with four children, using Macs and iPhones and Chromebooks. To have them send me "to do" requests would be so good!

**Reply by NaitLee** [reply]
I think for such small cases it's *nearly* enough now -- you can expose the Web UI (and in fact IPP service) to local network.
```bash
python3 server.py -a -s
```
Be sure to make its IP address static. On all family devices, just bookmark it.

For the cases you didn't mean rich text or whatever, right?
Nevertheless, one can edit in LibreOffice, snapshot the result & print as image.

Btw #31 is ongoing, the text printing on Web interface.
If you need it for trying out setting your env up, go for it -- just clone that fork.

I don't want to be rough either, so I think in short future there should be optional auth/verification there, at least.
And tell me if you have other "security" thoughts. One of mine is paper statistics & quota 🙃

----

**What if just make these enough in Web UI / cmdline backend?** 🤔

Stuffs like CUPS are not welcomed in Unix Philosophy because all the bloat in concept.
That's also a reason for the vendors not supporting them -- Neither devs nor users feel good about it.
